File: benchmarks/longmemeval/run_longmemeval_retrieval.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';

const args = process.argv.slice(2);
function arg(name, fallback) {
  const i = args.indexOf(name);
  return i >= 0 && i + 1 < args.length ? args[i + 1] : fallback;
}
const dataPath = arg('--data', 'benchmarks/longmemeval/data/longmemeval_oracle.json');
const outPath = arg('--out', 'benchmarks/longmemeval/results/oracle_retrieval_results.json');
const userOnly = args.includes('--user-only');

function tokens(text) {
  return String(text || '').toLowerCase().match(/[a-z0-9_]+/g)?.filter(t => t.length > 1) || [];
}

function sessionText(session) {
  return session
    .filter(turn => !userOnly || turn.role === 'user')
    .map(turn => `${turn.role || ''}: ${turn.content || ''}`)
    .join('\n');
}

function scoreSessions(question, sessions) {
  const qTokens = tokens(question);
  const docs = sessions.map(s => tokens(sessionText(s)));
  const df = new Map();
  for (const doc of docs) {
    for (const t of new Set(doc)) df.set(t, (df.get(t) || 0) + 1);
  }
  const nDocs = Math.max(1, docs.length);
  const avgdl = docs.reduce((a, d) => a + d.length, 0) / nDocs || 1;
  const qCounts = new Map();
  for (const t of qTokens) qCounts.set(t, (qCounts.get(t) || 0) + 1);
  return docs.map((doc, i) => {
    const counts = new Map();
    for (const t of doc) counts.set(t, (counts.get(t) || 0) + 1);
    const dl = Math.max(1, doc.length);
    let score = 0;
    for (const [t, qtf] of qCounts.entries()) {
      const tf = counts.get(t) || 0;
      if (!tf) continue;
      const dft = df.get(t) || 0;
      const idf = Math.log(1 + (nDocs - dft + 0.5) / (dft + 0.5));
      const denom = tf + 1.5 * (1 - 0.75 + 0.75 * dl / Math.max(avgdl, 1));
      score += idf * (tf * 2.5 / denom) * Math.min(qtf, 3);
    }
    return [i, score];
  }).sort((a, b) => b[1] - a[1]);
}

function reciprocalRank(rankedIds, answerIds) {
  for (let i = 0; i < rankedIds.length; i += 1) {
    if (answerIds.has(rankedIds[i])) return 1 / (i + 1);
  }
  return 0;
}

const data = JSON.parse(fs.readFileSync(dataPath, 'utf8'));
const metrics = { 'hit@1': 0, 'hit@3': 0, 'hit@5': 0, 'hit@10': 0, mrr: 0 };
const byType = new Map();
const rows = [];
let evaluated = 0;

for (const item of data) {
  if (String(item.question_id || '').endsWith('_abs')) continue;
  const answerIds = new Set(item.answer_session_ids || []);
  if (!answerIds.size) continue;
  const ranked = scoreSessions(item.question, item.haystack_sessions || []);
  const rankedIds = ranked.map(([i]) => item.haystack_session_ids[i]);
  const rr = reciprocalRank(rankedIds, answerIds);
  evaluated += 1;
  const qtype = item.question_type || 'unknown';
  if (!byType.has(qtype)) byType.set(qtype, { n: 0, 'hit@1': 0, 'hit@3': 0, 'hit@5': 0, 'hit@10': 0, mrr: 0 });
  const b = byType.get(qtype);
  b.n += 1;
  for (const k of [1, 3, 5, 10]) {
    const hit = rankedIds.slice(0, k).some(id => answerIds.has(id)) ? 1 : 0;
    metrics[`hit@${k}`] += hit;
    b[`hit@${k}`] += hit;
  }
  metrics.mrr += rr;
  b.mrr += rr;
  rows.push({ question_id: item.question_id, question_type: qtype, answer_session_ids: [...answerIds].sort(), top10: rankedIds.slice(0, 10), rr });
}

const summary = { dataset: dataPath, cases: evaluated, skipped_abstention_or_no_answer: data.length - evaluated };
for (const [k, v] of Object.entries(metrics)) summary[k] = evaluated ? Number((v / evaluated).toFixed(4)) : 0;
summary.by_type = {};
for (const [qtype, vals] of [...byType.entries()].sort()) {
  const n = vals.n;
  summary.by_type[qtype] = { n };
  for (const k of ['hit@1', 'hit@3', 'hit@5', 'hit@10', 'mrr']) summary.by_type[qtype][k] = Number((vals[k] / n).toFixed(4));
}

fs.mkdirSync(path.dirname(outPath), { recursive: true });
fs.writeFileSync(outPath, JSON.stringify({ summary, rows }, null, 2));
console.log(JSON.stringify(summary, null, 2));

File: benchmarks/longmemeval/run_longmemeval_retrieval.py
#!/usr/bin/env python3
"""Local LongMemEval retrieval benchmark for CausaMem-style memory retrieval.

This runner is intentionally dependency-light. It evaluates whether the memory
retrieval layer can bring the evidence session(s) into the top-k context.
"""

from __future__ import annotations

import argparse
import json
import math
import re
from collections import Counter, defaultdict
from pathlib import Path


TOKEN_RE = re.compile(r"[a-zA-Z0-9_]+")


def tokens(text: str) -> list[str]:
    return [t.lower() for t in TOKEN_RE.findall(text or "") if len(t) > 1]


def session_text(session: list[dict], user_only: bool = False) -> str:
    parts = []
    for turn in session:
        if user_only and turn.get("role") != "user":
            continue
        parts.append(f"{turn.get('role', '')}: {turn.get('content', '')}")
    return "\n".join(parts)


def score_sessions(question: str, sessions: list[list[dict]], user_only: bool = False) -> list[tuple[int, float]]:
    q_tokens = tokens(question)
    if not q_tokens:
        return [(i, 0.0) for i in range(len(sessions))]

    docs = [tokens(session_text(s, user_only=user_only)) for s in sessions]
    df = defaultdict(int)
    for doc in docs:
        for t in set(doc):
            df[t] += 1
    n_docs = max(1, len(docs))
    avgdl = sum(len(d) for d in docs) / n_docs if docs else 1.0
    q_counts = Counter(q_tokens)

    ranked = []
    for i, doc in enumerate(docs):
        counts = Counter(doc)
        dl = max(1, len(doc))
        score = 0.0
        for t, qtf in q_counts.items():
            tf = counts.get(t, 0)
            if not tf:
                continue
            idf = math.log(1 + (n_docs - df[t] + 0.5) / (df[t] + 0.5))
            denom = tf + 1.5 * (1 - 0.75 + 0.75 * dl / max(avgdl, 1.0))
            score += idf * (tf * 2.5 / denom) * min(qtf, 3)
        ranked.append((i, score))
    ranked.sort(key=lambda x: x[1], reverse=True)
    return ranked


def reciprocal_rank(ranked_ids: list[str], answer_ids: set[str]) -> float:
    for rank, sid in enumerate(ranked_ids, 1):
        if sid in answer_ids:
            return 1.0 / rank
    return 0.0


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--data", default="benchmarks/longmemeval/data/longmemeval_oracle.json")
    ap.add_argument("--out", default="benchmarks/longmemeval/results/oracle_retrieval_results.json")
    ap.add_argument("--user-only", action="store_true")
    args = ap.parse_args()

    data_path = Path(args.data)
    out_path = Path(args.out)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    data = json.loads(data_path.read_text(encoding="utf-8"))

    metrics = {"hit@1": 0, "hit@3": 0, "hit@5": 0, "hit@10": 0, "mrr": 0.0}
    by_type = defaultdict(lambda: {"n": 0, "hit@1": 0, "hit@3": 0, "hit@5": 0, "hit@10": 0, "mrr": 0.0})
    rows = []
    evaluated = 0

    for item in data:
        qid = item["question_id"]
        if qid.endswith("_abs"):
            continue
        answer_ids = set(item.get("answer_session_ids") or [])
        if not answer_ids:
            continue
        session_ids = item["haystack_session_ids"]
        ranked = score_sessions(item["question"], item["haystack_sessions"], user_only=args.user_only)
        ranked_ids = [session_ids[i] for i, _score in ranked]
        rr = reciprocal_rank(ranked_ids, answer_ids)
        evaluated += 1
        qtype = item.get("question_type", "unknown")
        by_type[qtype]["n"] += 1
        for k in (1, 3, 5, 10):
            hit = int(any(sid in answer_ids for sid in ranked_ids[:k]))
            metrics[f"hit@{k}"] += hit
            by_type[qtype][f"hit@{k}"] += hit
        metrics["mrr"] += rr
        by_type[qtype]["mrr"] += rr
        rows.append({
            "question_id": qid,
            "question_type": qtype,
            "answer_session_ids": sorted(answer_ids),
            "top10": ranked_ids[:10],
            "rr": rr,
        })

    summary = {"dataset": str(data_path), "cases": evaluated, "skipped_abstention_or_no_answer": len(data) - evaluated}
    for key, val in metrics.items():
        summary[key] = round(val / evaluated, 4) if evaluated else 0.0
    summary["by_type"] = {}
    for qtype, vals in sorted(by_type.items()):
        n = vals.pop("n")
        summary["by_type"][qtype] = {"n": n, **{k: round(v / n, 4) for k, v in vals.items()}}

    out_path.write_text(json.dumps({"summary": summary, "rows": rows}, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps(summary, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
